### ISSUE TITLE
BM25 returns no results in some situations (Original title: BM25/Tokenizer not working properly)

### ISSUE THREAD

**Issue by netapy** [weaviate/weaviate#3517]
Hi –

I have an **Article** class, which contains many french legal articles.

Here is the class initiation : 
```
class_obj = {
    "class": "Article",
    "description": "Articles des différentes codes de la loi.",
    "invertedIndexConfig": {
        "stopwords": {
            "preset": "none",
            #"additions": stopwords_fr
        }
    },
    "vectorizer": "text2vec-transformers",
    "properties": [
        {
            "name": "article_id",
            "description": "Id unique de l'article.",
            "dataType": [
                "text"
            ],
            "indexFilterable": True,
            "indexSearchable": False, 
        },
        {
            "name": "source",
            "description": "Titre de la source juridique (code, loi ou ordonnance) contenant l'article.",
            "dataType": [
                "text"
            ],
            "tokenization": "lowercase",
            "indexFilterable": True,
            "indexSearchable": True, 
        },
        {
            "name": "titre",
            "description": "Le titre de l'article.",
            "dataType": [
                "text"
            ],
            "indexFilterable": True,
            "indexSearchable": True, 
            "tokenization": "lowercase"
        },
        {
            "name": "texte",
            "description": "Le texte de l'article, en html.",
            "dataType": [
                "text"
            ],
            "moduleConfig": {
                "text2vec-transformers": {
                    "skip": False,
                    "vectorizePropertyName": False
                }
            },
            "indexFilterable": True,
            "indexSearchable": True, 
        },
        {
            "name": "etat",
            "description": "Etat de l'article : en vigueur, abrogé...",
            "dataType": [
                "text"
            ],
            "indexFilterable": True,
            "indexSearchable": False, 
        },
        {
            "name": "path_title",
            "description": "Chemin daccès à l'article",
            "dataType": [
                "text[]"
            ],
            "indexFilterable": True,
            "indexSearchable": True, 
        },
        {
            "name": "ref_textes",
            "description": "Références avec d'autres textes.",
            "dataType": [
                "text"
            ],
            "indexFilterable": True,
            "indexSearchable": False, 
        },
        {
            "name": "order",
            "description": "Ordre de l'article dans le code.",
            "dataType": [
                "int"
            ],
            "indexFilterable": True,
            "indexSearchable": False,
        },
        {
            "name": "date_deb",
            "description": "Date de début de l'article.",
            "dataType": [
                "int"
            ],
            "indexFilterable": True,
            "indexSearchable": False,
        },
        {
            "name": "date_fin",
            "description": "Date de fin de l'article.",
            "dataType": [
                "int"
            ],
            "indexFilterable": True,
            "indexSearchable": False,
        },
    ]
}

client.schema.create_class(class_obj)
```

I want to let my users search for legal articles through their "titre" and "source" properties, which have been tokenized using `lowercase`.

Here is an example of an article I'm trying to find :
```
{
    "article_id": "JORFARTI000047663197",
    "etat": "VIGUEUR",
    "path_title": [
        "Titre IER : DE LA NATURE DE L'ACTIVITÉ D'INFLUENCE COMMERCIALE PAR VOIE ÉLECTRONIQUE ET DES OBLIGATIONS AFFÉRENTES À SON EXERCICE",
        "Chapitre III : Dispositions générales relatives à l'activité d'agent d'influenceur, aux contrats d'influence commerciale par voie électronique, à la responsabilité civile solidaire et à l'assurance civile professionnelle"
    ],
    "source": "LOI n° 2023-451 du 9 juin 2023 visant à encadrer l'influence commerciale et à lutter contre les dérives des influenceurs sur les réseaux sociaux (1)",
    "texte": ".....blablabla....",
    "titre": "7"
},
```

Using the following query : 
```
query {
    Get {
    Article(
        limit: 5,
        bm25: {
            query: "LOI 9 juin 2023 visant à encadrer l'influence"
            properties: ["source^3", "titre"]
        }
        ) {
            article_id
            titre
            path_title
            texte
            source
            _additional {
                score
            }                
        }
    }
}
```

...gives me the following results : 
```
# only printing the "titre" and "source" in a list
[['2023', 'Code civil'], ["Annexe 9 à l'article A4241-50-2", 'Code des transports'], ["Annexe à l'article R*351-1, art. 9", 'Code des ports maritimes'], ['437 à 614-26', 'Code de commerce (ancien)'], ['L79 à L85', 'Code électoral']]
```

Am i missing something ? 
It seems like it should be able to find it because i'm literally copy/pasting the exact source name.
Could it be an issue with the lowercase tokenizer ? 

I'd be happy to provide you with further information if needed.

Thanks in advance.

**Comment by donomii** [comment]
Could you do the search with only the source property (remove the titre)?

e.g.

query {
    Get {
    Article(
        limit: 5,
        bm25: {
            query: "LOI 9 juin 2023 visant à encadrer l'influence"
            properties: ["source"]
        }
        ) {
            article_id
            titre
            path_title
            texte
            source
            _additional {
                score
            }                
        }
    }
}

**Comment by donomii** [comment]
Also, if you raise the limit (e.g. to 100), do the correct results appear in the list?

**Comment by netapy** [comment]
@donomii Thanks for your answer.

>Could you do the search with only the source property (remove the titre)?
I just tried, and weirdly enough it just gives me 0 results – doesn't find anything (out of +300k articles)

> Also, if you raise the limit (e.g. to 100), do the correct results appear in the list?
It doesn't show up.


It really seems that the *source* property has not been tokenized. 

I might try with the default "word" tokenization later to see if it solves it. Will keep this updated

**Comment by netapy** [comment]
Also, for what it's worth, when I batch add my *Articles* I directly provide the vector.

**Comment by netapy** [comment]
> I might try with the default "word" tokenization later to see if it solves it. Will keep this updated

Doing this doesn't change anything – *source* is still not tokenized.

**Comment by donomii** [comment]
Is this a public dataset?  Would I be able to download it and try to import it the same way you did?

Otherwise, could you give me some details about how much data, how many records, and the maximum size of the records?

**Comment by netapy** [comment]
> Is this a public dataset? Would I be able to download it and try to import it the same way you did?
> 
> Otherwise, could you give me some details about how much data, how many records, and the maximum size of the records?

No it is public data but unfortunately it needs quite a bit of structuring.
Would it help if I sent you my docker volume with all the articles ? 
Maybe over the weaviate slack

**Comment by MHilhorst** [comment]
we are having the same issues. Bm25 does not return any results while the objects do exist and we are querying on the correct class. The problem seem to occur after some time when uploading a new object. When a fresh object is uploaded bm25 seem to work for some time.

**Comment by donomii** [comment]
A quick update - we have confirmed this bug and are still investigating it

**Comment by narayanacharya6** [comment]
Can any details of the bug be shared and any possible workarounds (other than rewriting the data to Weaviate)?

**Comment by makarsergeev** [comment]
+1

**Comment by PythonAlchemist** [comment]
+1

**Comment by algebrei** [comment]
+1

**Comment by pavelnemirovsky** [comment]
+1

**Comment by donomii** [comment]
> Can any details of the bug be shared and any possible workarounds (other than rewriting the data to Weaviate)?

The bug appears to be index corruption, so your data is safe.  We could add an option to re-index the data inside weaviate while we fix the bug.

**Comment by narayanacharya6** [comment]
Thanks @donomii! 

1. How frequently will reindexing be done? Does this impact performance for other requests being served by Weaviate?
2. Can this be released as a patch?

**Comment by donomii** [comment]
> Thanks @donomii!
> 
> 1. How frequently will reindexing be done? Does this impact performance for other requests being served by Weaviate?
> 2. Can this be released as a patch?

1. Reindexing would be manual, and would certainly impact speed.
2. At this point I don't know how we would release it.

**Comment by narayanacharya6** [comment]
Understood. This impacts our adoption of Weaviate because we'd like to use the changes for `ContainsAny`/`ContainsAll` operators in v1.21 but BM25 and in-turn hybrid search is unusable in these versions. Please keep us posted with when a fix or option to manually re-index is available.

**Comment by krskrs** [comment]
What sounds a bit weird to me is how such a central feature has not been fixed yet. Maybe is everybody using weaviate only as a vector db? We had chosen it over its competitors just for its bm25 and hybrid support.

**Comment by alex-cortenix** [comment]
> What sounds a bit weird to me is how such a central feature has not been fixed yet. Maybe is everybody using weaviate only as a vector db? We had chosen it over its competitors just for its bm25 and hybrid support.

I am second to that, the main factor to choose Weaviate against competitors was a hybrid search. Seems the team was very busy releasing a new Python client (which BTW is irrelevant for us) and now should have more resources to fix this annoying issue...

**Comment by krskrs** [comment]
Hi @etiennedi that's great news! Unfortunately I haven't the time at this exact moment to test the script, but I can indeed  confirm your points.

- we started noticing this bug after upgrading to 1.21
- I was suspecting that it wasn't happening directly on write, but with some sort of "offline/delayed" operation, because our current workflow consists in importing some data, updating it from time to time (we do it, so no updates from users), and querying it for most part of the time. And I had noticed that suddenly, it would just stop working, but not immediately after the writes. We don't work with local weaviate instances, but on weaviate cloud only, so the flush interval is the default used on such instances.
